torch/fx/graph.py

```diff
@@ -487,6 +487,41 @@ def type_repr(o : Any):
                     type_repr(sub_type)
             return typename
 
+
+        # Run through reverse nodes and record the first instance of a use
+        # of a given node. This represents the *last* use of the node in the
+        # execution order of the program, which we will use to free unused
+        # values
+        node_to_last_use : Dict[Node, Node] = {}
+
+        def register_last_uses(n : Node, user : Node):
+            if n not in node_to_last_use:
+                node_to_last_use[n] = user
+
+        for node in reversed(self.nodes):
+            map_arg(node.args, lambda n: register_last_uses(n, node))
+            map_arg(node.kwargs, lambda n: register_last_uses(n, node))
+
+        def delete_unused_values(user : Node):
+            """
+            Delete values after their last use. This ensures that values that are
+            not used in the remainder of the code are freed and the memory usage
+            of the code is optimal.
+            """
+            nodes_to_delete : Set[Node] = set()
```

Comment:
sets are unordered, so this creates non-determinisitic code. making this a dictionary instead would make it deterministic.

Reply:
it's also possible to just construct this set in the `register_last_uses` function by maintaining a user -> last_uses map as it goes andd appending nodes as they are found. Its minor but it avoids having to scan the argument list of a node twice.